File: its-my-training-day/training-grounds/leetcode-top-75/questions/arrays-hashing/heap/top-k-frequent/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def top_k_frequent(nums: List[int], k: int) -> List[int]:
    freq: Dict[int, int] = {}
    i = 0
    while i < len(nums):
        val = nums[i]
        freq[val] = freq.get(val, 0) + 1
        i += 1
    heap = MinHeap()
    for key in freq:
        heap.push((freq[key], key))
        if heap.size() > k:
            heap.pop()
    result = [0] * k
    i = k - 1
    while i >= 0:
        result[i] = heap.pop()[1]
        i -= 1
    return result

File: its-my-training-day/training-grounds/leetcode-top-75/questions/arrays-hashing/heap/top-k-frequent/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

std::vector<int> topKFrequent(const std::vector<int>& nums, int k) {
    std::unordered_map<int, int> freq;
    for (int val : nums) freq[val] += 1;
    struct Pair {
        int count;
        int value;
        bool operator<(const Pair& other) const { return count < other.count; }
        bool operator>(const Pair& other) const { return count > other.count; }
    };
    MinHeap<Pair> heap;
    for (auto& kv : freq) {
        heap.push({kv.second, kv.first});
        if (static_cast<int>(heap.size()) > k) heap.pop();
    }
    std::vector<int> result(k);
    for (int i = k - 1; i >= 0; --i) {
        result[i] = heap.pop().value;
    }
    return result;
}
